blockindex/bloomfilterindexer.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2020 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"context"
+	"sync"
+
+	"github.com/iotexproject/go-pkgs/bloom"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/blockchain/blockdao"
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+	"github.com/pkg/errors"
+)
+
+const (
+	// BloomFilterNamespace indicates the kvstore namespace to store RangeBloomFilters
+	BloomFilterNamespace = "RangeBloomFilters"
+	// CurrentHeightKey indicates the key of current bf indexer height in underlying DB
+	CurrentHeightKey = "CurrentHeight"
+)
+
+type (
+	// BloomFilterIndexer is the interface for bloomfilter indexer
+	BloomFilterIndexer interface {
+		blockdao.BlockIndexer
+		RangeBloomFilterSize() uint64
+		RangeBloomFilter(uint64) (bloom.BloomFilter, error)
+	}
+
+	// bloomfilterIndexer is a struct for bloomfilter indexer
+	bloomfilterIndexer struct {
+		mutex               sync.RWMutex // mutex for curRangeBloomfilter
+		kvStore             db.KVStore
+		rangeSize           uint64
+		curRangeBloomfilter bloom.BloomFilter
+	}
+)
+
+// NewBloomfilterIndexer creates a new bloomfilterindexer struct by given kvstore and rangebloomfilter size
+func NewBloomfilterIndexer(kv db.KVStore, rangeSize uint64) (BloomFilterIndexer, error) {
+	if kv == nil {
+		return nil, errors.New("empty kvStore")
+	}
+	x := bloomfilterIndexer{
+		kvStore:   kv,
+		rangeSize: rangeSize,
+	}
+	return &x, nil
+}
+
+// Start starts the bloomfilter indexer
+func (bfx *bloomfilterIndexer) Start(ctx context.Context) (err error) {
+	if err := bfx.kvStore.Start(ctx); err != nil {
+		return err
+	}
+	bfx.mutex.Lock()
+	defer bfx.mutex.Unlock()
+	bfx.curRangeBloomfilter, err = bloom.NewBloomFilter(2048, 3)
+	if err != nil {
+		return err
+	}
```

Comment:
on start, it should be loaded from db if it is in the middle of a range

Reply:
Because it is an indexer, if it was in the middle of a range, when start(), it will check underlying DB tip Height which is smaller than dao.tipheight. And it will automatically catch up by putBlock() up to the blockdao's tip (it will regenerate the in-memory current bloomfilter).